_overviews/scala3-book/domain-modeling-fp.md

```diff
@@ -0,0 +1,408 @@
+
+
+
+## Modeling the data
+
+Modeling the “data” portion of a domain model in Scala is simple:
+
+- Create ADTs and GADTs with the `enum` construct
+- Use `case` classes for more complicated systems
+<!-- TODO: wording on that last line? -->
+
+
+### Crust size, crust type, and toppings
+
+Data concepts like crust size, crust type, and toppings are concisely modeled with the Scala 3 `enum` construct:
+
+```scala
+enum CrustSize:
+  case Small, Medium, Large
+
+enum CrustType:
+  case Thin, Thick, Regular
+
+enum Topping:
+  case Cheese, Pepperoni, BlackOlives, GreenOlives, Onions
+```
+
+
+### Modeling a pizza
+
+Now you can model a `Pizza` class using a `case` class:
+
+```scala
+import CrustSize._
+import CrustType._
+import Topping._
+
+case class Pizza(
+  crustSize: CrustSize,
+  crustType: CrustType,
+  toppings: Seq[Topping]
+)
+```
+
+And that’s it. That’s the data model for an FP-style pizza system. This solution is very concise because it doesn’t require the operations on a pizza to be combined with the data model. The data model is easy to read, like declaring the design for a relational database.
+
+<!-- TODO: compare this with C-style structs and others? -->
+
+#### More of the data model
+
+Because that was very short, you may prefer to see more of the data model for an entire pizza-ordering system. Here are a few other `case` classes that are used to model such a system:
+
+```scala
+case class Address(
+  street1: String,
+  street2: Option[String],
+  city: String,
+  state: String,
+  zipCode: String
+)
+
+case class Customer(
+  name: String,
+  phone: String,
+  address: Address
+)
+
+case class Order(
+  pizzas: Seq[Pizza],
+  customer: Customer
+)
+```
+
+#### “Skinny domain objects”
+
+In his book, *Functional and Reactive Domain Modeling*, Debasish Ghosh states that where OOP practitioners describe their classes as “rich domain models” that encapsulate data and behaviors, FP data models can be thought of as “skinny domain objects.” This is because — as this lesson shows — the data models are defined as `case` classes with attributes, but no behaviors.
+
+
+
+
+## What about the operations?
+
+This leads to an interesting question: Because FP separates the data from the operations on that data, how do you implement those operations in Scala/FP?
+
+There are several different ways to handle behaviors:
+
+- Put your methods in companion objects
+- Use a modular programming style
+- Use a “functional objects” approach
+
+These solutions are shown in the next three sections.
+<!-- TODO: is there ONE preferred solution? -->
+
+
+>An important point about all of the following solutions is that the methods and functions shown are *pure functions*. They don’t mutate any data; they receive an object, and apply transformation functions to that data to return new data.
+<!-- TODO: link to "pure function" definition -->
```

Comment:
The definition of pure function should more formal and precise

Reply:
I agree! :) Can you or anyone else recommend one? I haven’t looked at the comments further down on this Github page, so I don’t yet know if anyone address this, but I put a note about this in the *fp-pure-functions.md* file, looking for a better definition.

Reply:
You can draw inspiration from https://en.wikipedia.org/wiki/Functional_programming#Pure_functions (or just link to that page)